mixins/visibility-mixin.js

```diff
@@ -0,0 +1,207 @@
+			// done visibility transition, element is in original, fully visible state, so return the original transition to this
+			this.style.transition = this.transitionOriginal;
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-show-end',
+				{ bubbles: true, composed: false }
+			));
+		};
+		const animateShowStyle = {
+			initial: {
+				transition: `all ${transitionDuration}ms ease ${transitionDuration / 3}ms`,
+				opacity: '0',
+				transform: `translateY(-${moveYValue}px)`
+			},
+			initialDummy: {
+				transition: `height ${transitionDuration}ms ease`,
+				height: '0'
+			},
+			final: {
+				opacity: `${this.opacityOriginal}`,
+				transform: `${this.transformOriginal}`
+			},
+			finalDummy: {
+				height: `calc(${this.scrollHeight}px + ${window.getComputedStyle(this).marginTop} + ${window.getComputedStyle(this).marginBottom})`
+			},
+			dummyOnTransitionStart: dummyOnTransitionStart,
+			thisOnTransitionEnd: thisOnTransitionEnd
+		};
+
+		this._animateVisibility(animateShowStyle);
+	}
+
+	async _animateVisibility(animateStyle) {
+		if (!reduceMotion) {
+			Object.assign(this.style, animateStyle.initial);
+			Object.assign(this.dummy.style, animateStyle.initialDummy);
+
+			// we are in the middle of an earlier transition
+			if (document.body.contains(this.dummy)) {
+				// preserve the current opacity & transform for when we switch directions of transition
+				this.style.opacity = window.getComputedStyle(this).opacity;
+				this.style.transform = window.getComputedStyle(this).transform;
+
+				// preserve the current dummy height for when we switch directions of transition
+				this.dummy.style.height = window.getComputedStyle(this.dummy).height;
+				this.dummy.replaceWith(this);
+			}
+			this.replaceWith(this.dummy);
+			this.dummy.appendChild(this);
+
+			// allow enough time for reflow to occur to ensure that the transition properly runs
+			await new Promise((r) => requestAnimationFrame(() => requestAnimationFrame(r)));
+
+			this.ontransitionstart = (event) => {
+				if (event.target === this && event.propertyName === 'opacity') {
+					if (animateStyle.thisOnTransitionStart) {
+						animateStyle.thisOnTransitionStart();
+					}
+				}
+			};
+
+			this.dummy.ontransitionstart = (event) => {
+				if (event.target === this.dummy) {
+					if (animateStyle.dummyOnTransitionStart) {
+						animateStyle.dummyOnTransitionStart();
+					}
+				}
+			};
+
+			this.ontransitionend = (event) => {
+				if (event.target === this && event.propertyName === 'opacity') {
+					if (animateStyle.thisOnTransitionEnd) {
+						animateStyle.thisOnTransitionEnd();
+					}
+				}
+			};
+
+			this.dummy.ontransitionend = (event) => {
+				if (event.target === this.dummy) {
+					if (animateStyle.dummyOnTransitionEnd) {
+						animateStyle.dummyOnTransitionEnd();
+					}
+				}
+			};
+
+			Object.assign(this.dummy.style, animateStyle.finalDummy);
+			Object.assign(this.style, animateStyle.final);
+		}
+
```

Comment:
Can these move up to the top and then you can return after them and remove the nested `if`-statement?

```js
if (reduceMotion) {
  if (animateStyle.thisOnTransitionEnd ) animateStyle.thisOnTransitionEnd();
  if (animateStyle. dummyOnTransitionEnd ) animateStyle. dummyOnTransitionEnd();
  return;
}
// non-reduce motion code here
```

Reply:
Yeah this looks way better. I also ended up moving all `reduceMotion` code out of `_animateVisibility` and directly into the top of each of show/hide/remove instead.